#DPATCHLEVEL=0
# Correctly free old lines
--- src/libxosd/xosd.c~	2004-08-13 21:55:38.000000000 +0200
+++ src/libxosd/xosd.c	2004-08-16 15:24:22.000000000 +0200
@@ -866,6 +866,7 @@ int
 xosd_display(xosd * osd, int line, xosd_command command, ...)
 {
   int ret = -1;
+  union xosd_line newline = { type: LINE_blank };
   va_list a;
 
   FUNCTION_START(Dfunction);
@@ -883,7 +884,7 @@ xosd_display(xosd * osd, int line, xosd_
   case XOSD_printf:
     {
       char buf[2000];
-      struct xosd_text *l = &osd->lines[line].text;
+      struct xosd_text *l = &newline.text;
       char *string = va_arg(a, char *);
       if (command == XOSD_printf) {
         if (vsnprintf(buf, sizeof(buf), string, a) >= sizeof(buf)) {
@@ -895,19 +896,11 @@ xosd_display(xosd * osd, int line, xosd_
       if (string && *string) {
         ret = strlen(string);
         l->type = LINE_text;
-        if (l->string == NULL) {
-          l->string = malloc(ret + 1);
-        } else {
-          realloc(l->string, ret + 1);
-        }
+        l->string = malloc(ret + 1);
         memcpy(l->string, string, ret + 1);
       } else {
         ret = 0;
         l->type = LINE_blank;
-        if (l->string != NULL) {
-          free(l->string);
-          l->string = NULL;
-        }
         l->width = -1;
       }
       break;
@@ -916,7 +909,7 @@ xosd_display(xosd * osd, int line, xosd_
   case XOSD_percentage:
   case XOSD_slider:
     {
-      struct xosd_bar *l = &osd->lines[line].bar;
+      struct xosd_bar *l = &newline.bar;
       ret = va_arg(a, int);
       ret = (ret < 0) ? 0 : (ret > 100) ? 100 : ret;
       l->type = (command == XOSD_percentage) ? LINE_percentage : LINE_slider;
@@ -932,6 +925,16 @@ xosd_display(xosd * osd, int line, xosd_
   }
 
   _xosd_lock(osd);
+  /* Free old entry */
+  switch (osd->lines[line].type) {
+    case LINE_text:
+      free (osd->lines[line].text.string);
+    case LINE_blank:
+    case LINE_percentage:
+    case LINE_slider:
+      break;
+  }
+  osd->lines[line] = newline;
   osd->update |= UPD_content | UPD_timer | UPD_show;
   _xosd_unlock(osd);
 
