#DPATCHLEVEL=0
# fill_mask is no longer needed
--- src/libxosd/xosd.c~	2004-08-03 07:17:26.000000000 +0200
+++ src/libxosd/xosd.c	2004-08-03 07:18:08.000000000 +0200
@@ -150,8 +150,6 @@ struct xosd
 
   int timeout;                  /* delta time */
   struct timeval timeout_start; /* Absolute start of timeout */
-
-  enum { FILL_OUTLINE = 1, FILL_SHADOW = 2, FILL_FACE = 4 } fill_mask;
 };
 
 /** Global error string. */
@@ -348,23 +346,6 @@ expose_line(xosd * osd, int line)
   assert(osd && osd->fontset);
   FUNCTION_START(Dfunction);
 
-  osd->fill_mask = osd->outline_offset ? FILL_OUTLINE :
-    osd->shadow_offset ? FILL_SHADOW : FILL_FACE;
-
-  switch (osd->fill_mask) {
-  case FILL_FACE:
-    XSetForeground(osd->display, osd->gc, osd->pixel);
-    break;
-  case FILL_SHADOW:
-    XSetForeground(osd->display, osd->gc, osd->shadow_pixel);
-    break;
-  case FILL_OUTLINE:
-    XSetForeground(osd->display, osd->gc, osd->outline_pixel);
-    break;
-  }
-  XFillRectangle(osd->display, osd->line_bitmap, osd->gc,
-      0, y, osd->screen_width, osd->line_height);
-
   // Clear the XShape mask
   XFillRectangle(osd->display, osd->mask_bitmap, osd->mask_gc_back,
                  0, y, osd->screen_width, osd->line_height);
