#DPATCHLEVEL=0
# Remove xosd.length
--- src/libxosd/xosd.c~	2004-07-29 23:02:01.000000000 +0200
+++ src/libxosd/xosd.c	2004-07-30 17:03:13.000000000 +0200
@@ -91,7 +91,6 @@ typedef struct
   enum { LINE_blank, LINE_text, LINE_percentage, LINE_slider } type;
 
   char *text;
-  int length;
   int width;
   int percentage;
 } xosd_line;
@@ -263,15 +262,16 @@ static void
 draw_with_mask(xosd * osd, xosd_line * l, int inX, int inPlace, int inY,
                int draw_line_bitmap)
 {
+  int len = strlen(l->text);
   FUNCTION_START(Dfunction);
   XDRAWSTRING(osd->display,
               osd->mask_bitmap,
               osd->fontset,
-              osd->mask_gc, inX, inPlace + inY, l->text, l->length);
+              osd->mask_gc, inX, inPlace + inY, l->text, len);
   if (draw_line_bitmap) {
     XDRAWSTRING(osd->display,
                 osd->line_bitmap,
-                osd->fontset, osd->gc, inX, inY, l->text, l->length);
+                osd->fontset, osd->gc, inX, inY, l->text, len);
   }
 }
 
@@ -315,7 +315,7 @@ expose_line(xosd * osd, int line)
     break;
 
   case LINE_text:
-    if (!l->text || !l->length)
+    if (l->text == NULL)
       break;
     if (!osd->fontset) {
       DEBUG(Dtrace, "CRITICAL: No fontset");
@@ -511,8 +511,6 @@ timeout_loop(void *osdv)
 static int
 display_string(xosd * osd, xosd_line * l, char *string)
 {
-  XRectangle rect;
-
   assert(osd);
   FUNCTION_START(Dfunction);
   if (!osd->fontset) {
@@ -520,21 +518,27 @@ display_string(xosd * osd, xosd_line * l
     return -1;
   }
 
-  l->type = LINE_text;
-
-  if (string) {
-    l->length = strlen(string);
-    l->text = realloc(l->text, l->length + 1);
-    strcpy(l->text, string);
+  if (string && *string) {
+    XRectangle rect;
+    int len = strlen(string);
+    l->type = LINE_text;
+    if (l->text == NULL) {
+      l->text = strdup(string);
+    } else {
+      realloc(l->text, len + 1);
+      strcpy(l->text, string);
+    }
+    pthread_mutex_lock(&osd->mutex);
+    XmbTextExtents(osd->fontset, l->text, len, NULL, &rect);
+    l->width = rect.width;
+    pthread_mutex_unlock(&osd->mutex);
   } else {
-    l->text = realloc(l->text, 1);
-    l->text[0] = '\0';
-    l->length = 0;
+    l->type = LINE_blank;
+    if (l->text != NULL) {
+      free(l->text);
+      l->text = NULL;
+    }
   }
-  pthread_mutex_lock(&osd->mutex);
-  XmbTextExtents(osd->fontset, l->text, l->length, NULL, &rect);
-  pthread_mutex_unlock(&osd->mutex);
-  l->width = rect.width;
 
   return 0;
 }
@@ -651,7 +655,7 @@ set_font(xosd * osd, const char *font)
     if (l->type == LINE_text && l->text != NULL) {
       XRectangle rect;
 
-      XmbTextExtents(osd->fontset, l->text, l->length, NULL, &rect);
+      XmbTextExtents(osd->fontset, l->text, strlen(l->text), NULL, &rect);
       l->width = rect.width;
     }
   }
@@ -869,7 +873,7 @@ xosd_create(int number_lines)
   }
 
   for (i = 0; i < osd->number_lines; i++) {
-    osd->lines[i].type = LINE_text;
+    osd->lines[i].type = LINE_blank;
     osd->lines[i].text = NULL;
   }
 
