#DPATCHLEVEL=0
# x/y are use only in one place
--- src/libxosd/xosd.c.orig	2004-08-04 12:36:50.000000000 +0200
+++ src/libxosd/xosd.c	2004-08-04 12:45:39.000000000 +0200
@@ -124,8 +124,6 @@ struct xosd
   int screen_xpos;
   int height;
   int line_height;
-  int x;
-  int y;
   xosd_pos pos;
   xosd_align align;
   int hoffset;
@@ -1176,35 +1174,36 @@ xosd_set_font(xosd * osd, const char *fo
 static void
 update_pos(xosd * osd)
 {                               /* Requires mutex lock. */
+  int x = 0, y = 0;
   assert(osd);
   FUNCTION_START(Dfunction);
   switch (osd->pos) {
   case XOSD_bottom:
-    osd->y = osd->screen_height - osd->height - osd->voffset;
+    y = osd->screen_height - osd->height - osd->voffset;
     break;
   case XOSD_middle:
-    osd->y = osd->screen_height / 2 - osd->height - osd->voffset;
+    y = osd->screen_height / 2 - osd->height - osd->voffset;
     break;
   case XOSD_top:
   default:
-    osd->y = osd->voffset;
+    y = osd->voffset;
     break;
   }
 
   switch (osd->align) {
   case XOSD_left:
-    osd->x = osd->hoffset + osd->screen_xpos;
+    x = osd->hoffset + osd->screen_xpos;
     break;
   case XOSD_center:
-    osd->x = osd->hoffset + osd->screen_xpos;
+    x = osd->hoffset + osd->screen_xpos;
     /*
      * which direction should this default to, left or right offset 
      */
     break;
   case XOSD_right:
-    // osd->x = XDisplayWidth (osd->display, osd->screen) - osd->width 
+    // x = XDisplayWidth (osd->display, osd->screen) - osd->width 
     // - osd->hoffset; 
-    osd->x = -(osd->hoffset) + osd->screen_xpos;
+    x = -(osd->hoffset) + osd->screen_xpos;
     /*
      * neither of these work right, I want the offset to flip so
      * +offset is to the left instead of to the right when aligned
