#DPATCHLEVEL=0
# Set locale in osd_cat
--- src/osd_cat.c~	2004-08-26 19:37:13.000000000 +0200
+++ src/osd_cat.c	2004-08-26 19:37:04.000000000 +0200
@@ -14,6 +14,8 @@
 #include <unistd.h>
 #include <getopt.h>
 #include <errno.h>
+#include <locale.h>
+#include <X11/Xlib.h>
 #include <sys/time.h>
 
 static struct option long_options[] = {
@@ -66,6 +68,9 @@ xosd_align align = XOSD_left;
 int
 main(int argc, char *argv[])
 {
+  if (setlocale(LC_ALL, "") == NULL || !XSupportsLocale())
+    fprintf(stderr, "Locale not available, expect problems with fonts.\n");
+
   while (1) {
     int option_index = 0;
     int c =
