#DPATCHLEVEL=0
# Simplify the xosd_scroll
--- src/libxosd/xosd.c.orig	2004-08-04 13:31:06.000000000 +0200
+++ src/libxosd/xosd.c	2004-08-04 13:31:12.000000000 +0200
@@ -1595,42 +1595,29 @@ xosd_show(xosd * osd)
 int
 xosd_scroll(xosd * osd, int lines)
 {
-  int new_line;
+  int i;
+  xosd_line *src, *dst;
 
   FUNCTION_START(Dfunction);
   if (osd == NULL)
     return -1;
+  if (lines <= 0 || lines > osd->number_lines)
+    return -1;
 
   pthread_mutex_lock(&osd->mutex);
-  assert(lines > 0 && lines <= osd->number_lines);
-
-  /*
-   * First free everything no longer needed 
-   */
-  for (new_line = 0; new_line < lines; new_line++) {
-    if ((osd->lines[new_line].type == LINE_text)
-        && (osd->lines[new_line].text != NULL)) {
-      free(osd->lines[new_line].text);
-      osd->lines[new_line].text = NULL;
-      osd->lines[new_line].type = LINE_blank;
+  /* Clear old text */
+  for (i=0, src=osd->lines; i < lines; i++,src++)
+    if (src->text) {
+      free(src->text);
+      src->text = NULL;
     }
-  }
-
-  /*
-   * Do the scroll 
-   */
-  for (new_line = 0; new_line < osd->number_lines - lines; new_line++) {
-    osd->lines[new_line] = osd->lines[new_line + lines];
-  }
-
-  /*
-   * Clear the lines opened up by scrolling, need because of the use of
-   * realloc in display string 
-   */
-  while (new_line < osd->number_lines) {
-    osd->lines[new_line].text = NULL;
-    osd->lines[new_line].type = LINE_blank;
-    new_line++;
+  /* Move following lines forward */
+  for (dst=osd->lines; i < osd->number_lines; i++)
+    *dst++ = *src++;
+  /* Blank new lines */
+  for (;dst < src; dst++) {
+    dst->type = LINE_blank;
+    dst->text = NULL;
   }
   force_redraw(osd, -1);
   pthread_mutex_unlock(&osd->mutex);
